Add Action To Automatically Place Issues In Needs Triage Card If Appropriate

hamelsmu opened the pull request:
Summary of changes:

- Defined a spec for a GitHub Action in `/Issue_Triage/action`, which runs the code to triage issues as described in https://github.com/kubeflow/community/issues/278
- Minor changes to py/issue_triage/triage.py to allow for environment variables
- Addition of file `py/issue_triage/triage_for_action.py` as the entry point for the Action

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/kubeflow/code-intelligence/59)
<!-- Reviewable:end -->

hamelsmu commented:
@jlewi this is ready for review

/assign @jlewi

jlewi commented on .github/workflows/build_image.yaml:
```diff
@@ -0,0 +1,18 @@
+        run: |
+          cd $GITHUB_WORKSPACE
+          echo ${PASSWORD} | docker login -u $USERNAME --password-stdin
+          docker build -t hamelsmu/kubeflow-triage -f Issue_Triage/Dockerfile .
```
So this is currently pushing to your Docker registry? Could you add a TODO to change this to a Kubeflow GCR registry?

hamelsmu replied:
done

jlewi commented on py/code_intelligence/graphql.py:
```diff
@@ -9,10 +9,10 @@ class GraphQLClient(object):
   def __init__(self):
-    if not os.getenv("GITHUB_TOKEN"):
-      raise ValueError("GITHUB_TOKEN must be present as an environment variable upon instantiating this object.")
+    if not os.getenv("INPUT_PERSONAL_ACCESS_TOKEN"):
```
Why do we need to change the environment variable? Is it not possible to set the environment variable in the workflow?

hamelsmu replied:
The problem is that `GITHUB_TOKEN` is a special environment variable which you cannot override in Actions

jlewi commented on py/issue_triage/triage.py:
```diff
@@ -12,7 +12,15 @@
 import json
 
-TOKEN_NAME = "GITHUB_TOKEN"
+PROJECT_CARD_ID = os.getenv('INPUT_PROJECT_CARD_ID')
```
Should we continue to supply MD2... as a default value e.g.

```
PROJECT_CARD_ID = os.getenv('INPUT_PROJECT_CARD_ID', MDEzOlByb2plY3RDb2x1bW41OTM0MzEz)
```

hamelsmu replied:
This default value is passed from the workflow into the environment variable, so I think its better to keep things like this.   The default value is specified here: https://github.com/kubeflow/triage-issues/blob/master/action.yml#L8

jlewi commented on py/issue_triage/triage.py:
```diff
@@ -790,5 +789,11 @@ def _add_triage_project(self, issue_info):
-
-  fire.Fire(IssueTriage)
+  
+  triager = IssueTriage()
```
Can we continue to use Fire and have multiple entry points? Does GitHub actions not allow us to specify the command line to enter?

hamelsmu replied:
Yes

jlewi commented on triage.yaml:
```diff
@@ -0,0 +1,17 @@
+    runs-on: ubuntu-latest
+    steps:
+      - name: Update Kanban
+        uses: machine-learning-apps/kubeflow-triage@master
```
Should this be changed to kubeflow/kubeflow-triage?
So a GitHubAction step "uses" always points at a GitHub repository not a docker image?

hamelsmu replied:
I updated this but some additional changes you should be aware of:

- I changed the name of the repo to https://github.com/kubeflow/triage-issues, because it doesn't make sense to have the name kubeflow in the name of the repo.  

- You can refer directly to the docker container, but the interface becomes opaque to end users.  I'll leave it as is, pointing to kubeflow/triage-issues and we can change it later if you want

jlewi commented on triage.yaml:
```diff
@@ -0,0 +1,17 @@
+      - name: Update Kanban
+        uses: machine-learning-apps/kubeflow-triage@master
+        env:
+          GITHUB_TOKEN: ${{ secrets.GITHUB_TOKEN }}
```
It looks like you are setting the environment variable GITHUB_TOKEN here. So can't we just set it to the secret containing the GITHUB_TOKEN to modify the kanban board? I.e why do we need to change the code to use a different environment variable then the name GITHUB_TOKEN?

hamelsmu replied:
No there is another variable called GITHUB_TOKEN that you cannot override in the Actions environment, this is why I had to change the name of the environment variable.  This diff is a bit old I updated the interface PTAL

hamelsmu commented:
@jlewi I suggest you make some commits over this PR updating the container to GCR.  Its better to do this now, rather than later?  Make sure you put the proper credentials in the secrets etc that allow you to do this

hamelsmu commented:
@jlewi Summary of changes that I made:

- left the `fire` entrypoint in your script and made a seperate `triage_for_actions.py` file that imports that and acts as the entrypoint for the action.  This file really doesn't do anything except for input validation.  
- Changed references to the new Action location `kubeflow/triage-issues` (I changed the name of the repo)
- Yes I had to change the name of the environment variable GITHUB_TOKEN to something else, because this cannot be overridden in Actions. 

I actually didn't appreciate that you cared about this script and thought is was an ad-hoc throw away exercise that's why I changed it so much the first time around.  My apologies, I left things in place and built the Actions around the code here instead.

hamelsmu commented on triage.yaml:
```diff
@@ -0,0 +1,16 @@
+        with:
+          PROJECT_CARD_ID: 'MDEzOlByb2plY3RDb2x1bW41OTM0MzEz'
+          ISSUE_NUMBER: ${{ github.event.issue.number }}
+          PERSONAL_ACCESS_TOKEN: ${{ secrets.triage_projects_github_token }}
```
@jlewi You might be wondering why the parameter is  `PERSONAL_ACCESS_TOKEN` but in the python file I'm checking for `INPUT_PERSONAL_ACCESS_TOKEN`.  From [the docs](https://help.github.com/en/actions/automating-your-workflow-with-github-actions/metadata-syntax-for-github-actions#inputs):

```
When you specify an input to an action in a workflow file or use a default input value, GitHub creates an environment variable for the input with the name INPUT_<VARIABLE_NAME>. The environment variable created converts input names to uppercase letters and replaces spaces with _ characters.
```

hamelsmu replied:
Note the new signature of the action

hamelsmu commented on .github/workflows/triage_issues.yaml:
```diff
@@ -9,7 +9,7 @@ jobs:
     steps:
       - name: Update Kanban
-        uses: kubeflow/triage-issues@master
+        uses: kubeflow/code-intelligence/Issue_Triage/Action@master
```
Note the new location of the Action

hamelsmu commented on .github/workflows/triage_issues.yaml:
```diff
@@ -9,7 +9,7 @@ jobs:
     steps:
       - name: Update Kanban
-        uses: kubeflow/triage-issues@master
+        uses: kubeflow/code-intelligence/Issue_Triage/action@master
```
Note the new location of the action

jlewi commented:
Regarding using GCR vs DockerHub. I agree it would be best if we could try to use GCR. This [action](https://github.com/actions/gcloud/tree/master/auth) suggests we can just login using gcloud. And then subsequent steps should be able to use the GCP service account (I'm guessing you already did something like this in your other examples).

I think we may need to run more step [docs](https://cloud.google.com/container-registry/docs/advanced-authentication#gcloud_as_a_docker_credential_helper)

```
gcloud auth configure-docker
```

I will create an appropriate service account and upload the key.

jlewi commented:
I have created the service account github-actions in project kubeflow-ci and uploaded it to the secret
github_actions_gcp_sa

This service account should be able to push images to gcr.io/kubeflow-ci

@hamelsmu Could we try using GCR instead of DockerHub?  So far I haven't found a really good way to manage DockerHub accounts; e.g. there's no notion of a service account that makes it easy to automate access like we can with GCP.

hamelsmu commented:
@jlewi pushing a container to gcr is not a problem the only problem is defining a “Docker Action” that other people can inherit in their workflows i can’t find a way to reference other registries in https://github.com/kubeflow/code-intelligence/blob/master/Issue_Triage/action/action.yml#L20

I’ll have to find an answer for this and get back to you

jlewi commented:
https://github.com/kubeflow/code-intelligence/commit/bebb446dee7e79f1a88f008ba13647af9a472263/checks?check_suite_id=328000249

It looks like 

* Its still using a docker image hamelsmu/kubeflow-triage
* The project id card isn't set so it error'd out.

It looks like it used the code on master: 
Download action repository 'kubeflow/code-intelligence@master'

Is there some way to change that so it uses this PR so we can use the code in this PR?

Commenting on a couple of open issues

## Repository customization

* For the purposes of Kubeflow I think hard coding the registry is fine; it would be the same across all Kubeflow repositories
* So if we treat the action as Kubeflow specific (and not some generic action) then hard coding the registry (at least for now)


## Github token

* Could we do the following

   * Look for the environment variable "GITHUB_TOKEN_ISSUE_TRIAGE" ; if its set use it
   * If "GITHUB_OTKEN_ISSUE_TRIAGE" isn't set try "GITHUB_TOKEN"
   * If neither is set raise error

* The motivation for falling back to GITHUB_TOKEN is that it looks like using that environment variable is a common convention
    * My expectation would be that people set that environment variable in their .bashrc or similar file so  that all scripts work locally
    * So it would be nice to respect that convention rather than force people to always set application specific tokens
   * Allowing that to be overwritten by application specific tokens seems beneficial for uses like GitHub actions.

hamelsmu commented:
> Is there some way to change that so it uses this PR so we can use the code in this PR?
 
Good catch.  I set the action to this branch instead of `@master`

> For the purposes of Kubeflow I think hard coding the registry is fine; it would be the same across all Kubeflow repositories

Keep in mind that we don't need a registry, we are only using a registry to "cache" the environment so it doesn't have to build each time we run the Action.  I'm assuming you care about the 5x speedup, only for the sake of conserving your free credits.  Just wanted to let you know.

  > If "GITHUB_OTKEN_ISSUE_TRIAGE" isn't set try "GITHUB_TOKEN"

Fixed this in https://github.com/kubeflow/code-intelligence/pull/59/commits/aea811146be6981b120e81c3229252c2258d8b08

jlewi commented:
It looks like the logs correspond to this command being invoked
https://github.com/kubeflow/code-intelligence/blob/ecbab8b8ca38ce8175c0a2550ab2a916e14d0e1d/py/issue_triage/triage.py#L682

I can repo this running locally so doesn't seem like an issue with this PR per se but rather the triage.py code.

jlewi commented:
Ok looks like the issue is that issue_triage.py doesn't take into account labels being removed when deciding whether an issue needs triage.

I opened a new issue.
kubeflow/code-intelligence#62

Looks like it worked successfully. This is really sweet!

jlewi commented:
@hamelsmu just a couple nitpicky comments about the token name; you can see inline comments but basically

1. Can we use "GITHUB_PRESONAL_ACCESS_TOKEN" inside graphql.py and triage.py
1. In the wrapper script for triage.py for GitHub actions; read environment variables with "INPUT_" prefix and set an environment variable of the same name but minus the "INPUT_" prefix
    * e.g set environment variable GITHUB_PERSONAL_ACCESS_TOKEN to the value of INPUT_GITHUB_PERSONAL_ACCESS_TOKEN

jlewi commented:
Woo Hoo. 

This looks great thank you so much @hamelsmu. 

/lgtm
/approve

I'll file a couple follow on issues about further improvements.
